{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The scikit-learn version is 1.2.1.\n"
     ]
    }
   ],
   "source": [
    "from hdf5storage import loadmat, savemat \n",
    "import numpy as np \n",
    "\n",
    "from scipy import signal \n",
    "from scipy.fftpack import fft\n",
    "\n",
    "from sklearn.covariance import GraphicalLassoCV\n",
    "\n",
    "import sklearn\n",
    "print('The scikit-learn version is {}.'.format(sklearn.__version__)) \n",
    "# make sure it is newer version, works in version 1.2.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load condition_all\n",
    "outdict=loadmat('conditions_all.mat')\n",
    "conditions_all=outdict[\"conditions_all\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Organize indicies \n",
    "Uncoupled_Ind=np.zeros((12,3))\n",
    "L_Lead_Ind=np.zeros((12,3)) # they were leading indicies for L subject but following for R\n",
    "R_Lead_Ind=np.zeros((12,3))\n",
    "Mutual_Ind=np.zeros((12,3))\n",
    "\n",
    "for ses in range(12):\n",
    "    Uncoupled_Ind[ses]=np.asarray(np.where(conditions_all[ses]==1))\n",
    "    L_Lead_Ind[ses]=np.asarray(np.where(conditions_all[ses]==2))\n",
    "    R_Lead_Ind[ses]=np.asarray(np.where(conditions_all[ses]==3))\n",
    "    Mutual_Ind[ses]=np.asarray(np.where(conditions_all[ses]==4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load abs_zscore_all\n",
    "outdict=loadmat('abs_zscore_all.mat')\n",
    "abs_zscore_all=outdict[\"abs_zscore_all\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# append abs_zscore into 4 conditions in 2 syn types\n",
    "# append all trials of the same condition together\n",
    "synch_Uncoupled = list()\n",
    "synch_Leading   = list()\n",
    "synch_Following = list()\n",
    "synch_Mutual    = list()\n",
    "\n",
    "synco_Uncoupled = list()\n",
    "synco_Leading   = list()\n",
    "synco_Following = list()\n",
    "synco_Mutual    = list()\n",
    "\n",
    "\n",
    "for ses in list(range(0,12,2)): # synch [0, 2, 4, 6, 8, 10]\n",
    "    for subj in range(2):\n",
    "        for trl in range(3):\n",
    "            if subj == 0:\n",
    "                synch_Uncoupled.append(abs_zscore_all[ses][int(Uncoupled_Ind[ses][trl])][subj])\n",
    "                synch_Leading.append(abs_zscore_all[ses][int(L_Lead_Ind[ses][trl])][subj])  \n",
    "                synch_Following.append(abs_zscore_all[ses][int(R_Lead_Ind[ses][trl])][subj])    \n",
    "                synch_Mutual .append(abs_zscore_all[ses][int(Mutual_Ind[ses][trl])][subj])  \n",
    "            else:\n",
    "                synch_Uncoupled.append(abs_zscore_all[ses][int(Uncoupled_Ind[ses][trl])][subj])\n",
    "                synch_Leading.append(abs_zscore_all[ses][int(R_Lead_Ind[ses][trl])][subj])  \n",
    "                synch_Following.append(abs_zscore_all[ses][int(L_Lead_Ind[ses][trl])][subj])   \n",
    "                synch_Mutual .append(abs_zscore_all[ses][int(Mutual_Ind[ses][trl])][subj])  \n",
    "\n",
    "for ses in list(range(1,12,2)): # [1, 3, 5, 7, 9, 11]\n",
    "    for subj in range(2):\n",
    "        for trl in range(3):\n",
    "            if subj == 0:\n",
    "                synco_Uncoupled.append(abs_zscore_all[ses][int(Uncoupled_Ind[ses][trl])][subj])\n",
    "                synco_Leading.append(abs_zscore_all[ses][int(L_Lead_Ind[ses][trl])][subj])  \n",
    "                synco_Following.append(abs_zscore_all[ses][int(R_Lead_Ind[ses][trl])][subj])    \n",
    "                synco_Mutual .append(abs_zscore_all[ses][int(Mutual_Ind[ses][trl])][subj])  \n",
    "            else:\n",
    "                synco_Uncoupled.append(abs_zscore_all[ses][int(Uncoupled_Ind[ses][trl])][subj])\n",
    "                synco_Leading.append(abs_zscore_all[ses][int(R_Lead_Ind[ses][trl])][subj])  \n",
    "                synco_Following.append(abs_zscore_all[ses][int(L_Lead_Ind[ses][trl])][subj])\n",
    "                synco_Mutual .append(abs_zscore_all[ses][int(Mutual_Ind[ses][trl])][subj])\n",
    "\n",
    "\n",
    "#  combine the 4 conditions each concatenated in a list\n",
    "synch_4condi=[np.concatenate(synch_Uncoupled), np.concatenate(synch_Leading), np.concatenate(synch_Following), np.concatenate(synch_Mutual)]\n",
    "synco_4condi=[np.concatenate(synco_Uncoupled), np.concatenate(synco_Leading), np.concatenate(synco_Following), np.concatenate(synco_Mutual)]\n",
    "syn_4condi=[synch_4condi, synco_4condi]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def partialcorrelation(ampdata,ncv=5):\n",
    "    pc_lasso = GraphicalLassoCV(cv=ncv)\n",
    "    nbin = np.shape(ampdata)[1]\n",
    "    nchan = np.shape(ampdata)[2]\n",
    "    pcorr = np.zeros((nbin,nchan,nchan))\n",
    "    alphas = np.zeros(nbin)\n",
    "    # likelihoods = list()\n",
    "    likelihoods = np.zeros(nbin)\n",
    "    covmat = np.zeros((nbin,nchan,nchan))\n",
    "    for f in range(nbin):\n",
    "        xx= np.abs(np.squeeze(ampdata[:,f,:]))\n",
    "        pc_lasso.fit(xx) # the system is too ill for this solver\n",
    "        covariance = pc_lasso.covariance_\n",
    "        precision = pc_lasso.precision_\n",
    "        alphas[f] = pc_lasso.alpha_\n",
    "        dict_cvresults=pc_lasso.cv_results_\n",
    "        best_alpha_ind=np.where(dict_cvresults['alphas']==pc_lasso.alpha_)\n",
    "        # likelihoods.append(pc_lasso.cv_results_[\"mean_test_score\"])\n",
    "        likelihoods[f] = pc_lasso.cv_results_[\"mean_test_score\"][best_alpha_ind][0]\n",
    "        parcor = np.zeros((32,32))\n",
    "        y = np.diag(precision)\n",
    "        yy = np.outer(y,y)\n",
    "        yy = np.sqrt(yy)\n",
    "        parcor = precision/yy\n",
    "        pcorr[f,:,:] = parcor\n",
    "        y = np.diag(covariance)\n",
    "        yy = np.outer(y,y)\n",
    "        yy = np.sqrt(yy)\n",
    "        parcor_cov = covariance/yy        \n",
    "        covmat[f,:,:] = parcor_cov \n",
    "    return pcorr,covmat,alphas,likelihoods  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parameters \n",
    "epoch = 1 #units: seconds\n",
    "maxf = 30 #units: Hz\n",
    "maxbin = maxf*epoch #convert maxf into number of bins to save"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:272: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[indices != idx, idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/sklearn/covariance/_graph_lasso.py:273: RuntimeWarning: invalid value encountered in multiply\n",
      "  precision_[idx, indices != idx] = -precision_[idx, idx] * coefs\n",
      "/home/zhibinz2/anaconda3/lib/python3.9/site-packages/numpy/core/_methods.py:230: RuntimeWarning: invalid value encountered in subtract\n",
      "  x = asanyarray(arr - arrmean)\n"
     ]
    }
   ],
   "source": [
    "# fit lasso model for each of the 4 conditions in 2 syn types\n",
    "synch_pcorr =   np.zeros((4,maxbin,32,32));       synco_pcorr  =np.zeros((4,maxbin,32,32))\n",
    "synch_covmat=   np.zeros((4,maxbin,32,32));       synco_covmat =np.zeros((4,maxbin,32,32))\n",
    "synch_alpha =   np.zeros((4,maxbin));             synco_alpha  =np.zeros((4,maxbin))\n",
    "synch_likelihood = np.zeros((4,maxbin)) #list()      \n",
    "synco_likelihood = np.zeros((4,maxbin)) #list()\n",
    "for condi in range(4):\n",
    "    pcorr,covmat,alphas,likelihoods = partialcorrelation(synch_4condi[condi],ncv=5)\n",
    "    synch_pcorr[condi,:,:,:]  =pcorr\n",
    "    synch_covmat[condi,:,:,:] =covmat\n",
    "    synch_alpha[condi,:]    =alphas\n",
    "    # synch_likelihood.append(likelihoods)\n",
    "    synch_likelihood[condi,:] =likelihoods\n",
    "    pcorr,covmat,alphas,likelihoods = partialcorrelation(synco_4condi[condi],ncv=5)\n",
    "    synco_pcorr [condi,:,:,:] =pcorr\n",
    "    synco_covmat[condi,:,:,:] =covmat\n",
    "    synco_alpha[condi,:]    =alphas\n",
    "    # synco_likelihood.append(likelihoods)\n",
    "    synco_likelihood[condi,:] =likelihoods\n",
    "outdict = dict()\n",
    "outdict['synch_pcorr'] = synch_pcorr\n",
    "outdict['synco_pcorr'] = synco_pcorr\n",
    "outdict['synch_covmat']= synch_covmat, \n",
    "outdict['synco_covmat'] = synco_covmat\n",
    "outdict['synch_alpha'] = synch_alpha\n",
    "outdict['synco_alpha'] = synco_alpha\n",
    "outdict['synch_likelihood'] = synch_likelihood\n",
    "outdict['synco_likelihood'] = synco_likelihood\n",
    "\n",
    "# takes 3 min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save syn_pcorr\n",
    "\n",
    "savemat('syn_pcorr',outdict,store_python_metadata=True)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "17b0b751a843f9cb4dea2611a4ce474c0b5eafa09e2c3e69ad3a32fe03d403fd"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
